File: go-cli/discovery.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

// GetSessionsDir returns the sessions directory, respecting env override.
func GetSessionsDir() string {
	if env := os.Getenv("COPILOT_SESSIONS_DIR"); env != "" {
		return env
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return filepath.Join(home, ".copilot", "session-state")
}

// ScanSessions scans a sessions directory and returns metadata sorted newest-first.
func ScanSessions(sessionsDir, repoFilter string, since *time.Time, search string) []WorkspaceMetadata {
	if sessionsDir == "" {
		sessionsDir = GetSessionsDir()
	}
	entries, err := os.ReadDir(sessionsDir)
	if err != nil {
		return nil
	}

	var sessions []WorkspaceMetadata
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		wsPath := filepath.Join(sessionsDir, entry.Name(), "workspace.yaml")
		data, err := os.ReadFile(wsPath)
		if err != nil {
			continue
		}
		var ws WorkspaceMetadata
		if err := yaml.Unmarshal(data, &ws); err != nil {
			continue
		}

		// Apply repo filter (case-insensitive glob match)
		if repoFilter != "" {
			if ws.Repository == "" {
				continue
			}
			matched, _ := filepath.Match(strings.ToLower(repoFilter), strings.ToLower(ws.Repository))
			if !matched {
				continue
			}
		}

		// Apply since filter
		if since != nil && ws.CreatedAt != nil && ws.CreatedAt.Before(*since) {
			continue
		}

		// Apply search filter (substring in summary)
		if search != "" {
			if !strings.Contains(strings.ToLower(ws.Summary), strings.ToLower(search)) {
				continue
			}
		}

		sessions = append(sessions, ws)
	}

	// Sort newest-first
	sort.Slice(sessions, func(i, j int) bool {
		ti := sessions[i].CreatedAt
		tj := sessions[j].CreatedAt
		if ti == nil && tj == nil {
			return false
		}
		if ti == nil {
			return false
		}
		if tj == nil {
			return true
		}
		return ti.After(*tj)
	})

	return sessions
}

// ResolveSession resolves a session specifier to a directory path.
func ResolveSession(specifier, sessionsDir string) (string, error) {
	if sessionsDir == "" {
		sessionsDir = GetSessionsDir()
	}

	// 1. Check if it's a direct path
	if info, err := os.Stat(specifier); err == nil && info.IsDir() {
		evPath := filepath.Join(specifier, "events.jsonl")
		if _, err := os.Stat(evPath); err == nil {
			return specifier, nil
		}
	}

	// 2. Try as index number
	if idx, err := strconv.Atoi(specifier); err == nil {
		sessions := ScanSessions(sessionsDir, "", nil, "")
		if idx >= 1 && idx <= len(sessions) {
			candidate := filepath.Join(sessionsDir, sessions[idx-1].ID)
			if info, err := os.Stat(candidate); err == nil && info.IsDir() {
				return candidate, nil
			}
		}
		return "", fmt.Errorf("index %d out of range (1–%d)", idx, len(sessions))
	}

	// 3. Try as full or partial UUID
	specLower := strings.ToLower(strings.TrimSpace(specifier))
	entries, err := os.ReadDir(sessionsDir)
	if err != nil {
		return "", fmt.Errorf("sessions directory not found: %s", sessionsDir)
	}

	var matches []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		name := entry.Name()
		nameLower := strings.ToLower(name)
		if nameLower == specLower {
			return filepath.Join(sessionsDir, name), nil // exact match
		}
		if strings.HasPrefix(nameLower, specLower) {
			matches = append(matches, name)
		}
	}

	if len(matches) == 1 {
		return filepath.Join(sessionsDir, matches[0]), nil
	}
	if len(matches) > 1 {
		shown := matches
		suffix := ""
		if len(shown) > 5 {
			shown = shown[:5]
			suffix = "..."
		}
		return "", fmt.Errorf("ambiguous prefix '%s' matches %d sessions: %s%s",
			specifier, len(matches), strings.Join(shown, ", "), suffix)
	}
	return "", fmt.Errorf("no session found matching '%s'", specifier)
}

File: src/copilot_export/discovery.py
"""Session discovery — scan, list, and resolve Copilot CLI sessions."""

from __future__ import annotations

import fnmatch
import os
from datetime import datetime
from pathlib import Path

import yaml

from .models import WorkspaceMetadata

DEFAULT_SESSIONS_DIR = Path.home() / ".copilot" / "session-state"


def get_sessions_dir() -> Path:
    """Return the sessions directory, respecting env var override."""
    env = os.environ.get("COPILOT_SESSIONS_DIR")
    if env:
        return Path(env)
    return DEFAULT_SESSIONS_DIR


def scan_sessions(
    sessions_dir: Path | None = None,
    repo_filter: str | None = None,
    since: datetime | None = None,
    search: str | None = None,
) -> list[WorkspaceMetadata]:
    """Scan a sessions directory and return metadata for all valid sessions.

    Results are sorted newest-first by created_at.
    """
    root = sessions_dir or get_sessions_dir()
    if not root.is_dir():
        return []

    sessions: list[WorkspaceMetadata] = []
    for entry in root.iterdir():
        if not entry.is_dir():
            continue
        ws_path = entry / "workspace.yaml"
        if not ws_path.exists():
            continue
        try:
            with open(ws_path, "r", encoding="utf-8") as f:
                data = yaml.safe_load(f) or {}
            ws = WorkspaceMetadata.from_dict(data)
        except Exception:
            continue

        # Apply filters
        if repo_filter and ws.repository:
            if not fnmatch.fnmatch(ws.repository.lower(), repo_filter.lower()):
                continue
        elif repo_filter and not ws.repository:
            continue

        if since and ws.created_at and ws.created_at < since:
            continue

        if search:
            searchable = (ws.summary or "").lower()
            if search.lower() not in searchable:
                continue

        sessions.append(ws)

    # Sort newest-first
    sessions.sort(key=lambda s: s.created_at or datetime.min, reverse=True)
    return sessions


def resolve_session(
    specifier: str,
    sessions_dir: Path | None = None,
) -> Path:
    """Resolve a session specifier to a directory path.

    Specifier can be:
    - A directory path (returned as-is if valid)
    - A full UUID
    - A partial UUID prefix (git-style, must be unique)
    - An index number (1-based, from newest-first scan)
    """
    root = sessions_dir or get_sessions_dir()

    # 1. Check if it's a direct path
    as_path = Path(specifier)
    if as_path.is_dir() and (as_path / "events.jsonl").exists():
        return as_path

    # 2. Try as index number
    try:
        idx = int(specifier)
        sessions = scan_sessions(sessions_dir=root)
        if 1 <= idx <= len(sessions):
            session_id = sessions[idx - 1].id
            candidate = root / session_id
            if candidate.is_dir():
                return candidate
        raise ValueError(
            f"Index {idx} out of range (1–{len(sessions)})"
        )
    except ValueError as e:
        if "out of range" in str(e):
            raise
        pass  # Not a number, try UUID matching

    # 3. Try as full or partial UUID
    specifier_lower = specifier.lower().strip()
    if not root.is_dir():
        raise FileNotFoundError(f"Sessions directory not found: {root}")

    matches: list[Path] = []
    for entry in root.iterdir():
        if not entry.is_dir():
            continue
        if entry.name.lower() == specifier_lower:
            return entry  # Exact match
        if entry.name.lower().startswith(specifier_lower):
            matches.append(entry)

    if len(matches) == 1:
        return matches[0]
    elif len(matches) > 1:
        match_ids = [m.name for m in matches]
        raise ValueError(
            f"Ambiguous prefix '{specifier}' matches {len(matches)} sessions: "
            + ", ".join(match_ids[:5])
            + ("..." if len(match_ids) > 5 else "")
        )
    else:
        raise FileNotFoundError(
            f"No session found matching '{specifier}'"
        )
